# Model Development

### Objetivos

Luego de completar este curso seras capas de:
 - Desarrollar modelos de predicción

En esta sección, desarrollaremos varios modelos que predeciran el precio del vehículo las variables o caracteristicas. Esto es solo un estimado pero nos debería dar una idea objetiva de cuando debería costar un vehículo.

Algunas preguntas para hacernos en este modulo

 - Se si el vendedor esta ofreciendo un precio justo por mi intercambio?
 - Se si he puesto el precio justo para mi vehículo?

En data analytics, normalmente usamos **Desarrollo de Modelos** para ayudarnos a predecir futuras observaciones de la información que tenemos.

Un modelo nos ayudara a entender la relación exacta entre diferentes variables y como estas variables son usadas para predecir el resultado.

Preparación

importar librerías:

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [2]:
filename = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/automobileEDA.csv"

df = pd.read_csv(filename, header=0)

df.head(5)

symboling  normalized-losses         make aspiration num-of-doors  \
0          3                122  alfa-romero        std          two   
1          3                122  alfa-romero        std          two   
2          1                122  alfa-romero        std          two   
3          2                164         audi        std         four   
4          2                164         audi        std         four   

    body-style drive-wheels engine-location  wheel-base    length  ...  \
0  convertible          rwd           front        88.6  0.811148  ...   
1  convertible          rwd           front        88.6  0.811148  ...   
2    hatchback          rwd           front        94.5  0.822681  ...   
3        sedan          fwd           front        99.8  0.848630  ...   
4        sedan          4wd           front        99.4  0.848630  ...   

   compression-ratio  horsepower  peak-rpm city-mpg highway-mpg    price  \
0                9.0       111.0    5000.0       21          27  13495.0   
1                9.0       111.0    5000.0       21          27  16500.0   
2                9.0       154.0    5000.0       19          26  16500.0   
3               10.0       102.0    5500.0       24          30  13950.0   
4                8.0       115.0    5500.0       18          22  17450.0   

  city-L/100km  horsepower-binned  diesel  gas  
0    11.190476             Medium       0    1  
1    11.190476             Medium       0    1  
2    12.368421             Medium       0    1  
3     9.791667             Medium       0    1  
4    13.055556             Medium       0    1  

[5 rows x 29 columns]

## 1. Regresión lineal y Regresión lineal multiple.

### Regresión lineal

Un ejemplo de modelo de datos que estaremos usando es:

**Regresión lineal simple**


La Regresión Lineal Simple es un método que nos ayuda a entender la relación entre dos variables:

 - La variable predictora/independiente (X)
 - La variable de respuesta/dependiente (que queremos predecir) (Y)
El resultado de la Regresión Lineal es una función lineal que predice la variable de respuesta (dependiente) como una función de la variable predictora (independiente).

$$
Y: Variable de Respuesta\\\\\\\\\\\\ 
X: Variables Predictoras
$$

**Función Lineal**
$$
Yhat = a + b  X
$$

 - a se refiere a la intersección de la línea de regresión, en otras palabras: el valor de Y cuando X es 0.

 - b se refiere a la pendiente de la línea de regresión, en otras palabras: el valor con el cual Y cambia cuando X aumenta en 1 unidad.

Vamos a importar los módulos para la Regresión Lineal:

In [7]:
from sklearn.linear_model import LinearRegression

#### Creamos el objeto Regresión lineal

In [8]:
lm = LinearRegression()
lm

LinearRegression()


¿Cómo podría "highway-mpg" ayudarnos a predecir el precio de un automóvil?
Para este ejemplo, queremos analizar cómo "highway-mpg" puede ayudarnos a predecir el precio del automóvil. Utilizando regresión lineal simple, crearemos una función lineal con "highway-mpg" como la variable predictora y el "precio" como la variable de respuesta.

In [10]:
X = df[["highway-mpg"]]
Y = df["price"]

Ajusta el modelo lineal utilizando "highway-mpg":

In [11]:
lm.fit(X,Y)

LinearRegression()

Podemos generar una predicción:

In [12]:
Yhat=lm.predict(X)
Yhat[0:5]

array([16236.50464347, 16236.50464347, 17058.23802179, 13771.3045085 ,
       20345.17153508])

#### ¿Cuál es el valor de la intersección (a)?

In [13]:
lm.intercept_

38423.3058581574

#### ¿Cuál es el valor de la pendiente (b)?

In [14]:
lm.coef_

array([-821.73337832])


¿Cuál es el modelo lineal estimado final que obtenemos?
Como vimos anteriormente, deberíamos obtener un modelo lineal final con la estructura:

$$
Yhat = a + b  X
$$

Al insertar los valores reales, obtenemos:

**Price** = 38423.31 - 821.73 x highway-mpg

# Pregunta #1 a)

**Crea un objeto regresión lineal llamado "lm1"**

In [24]:
lm1 = LinearRegression()
lm1

LinearRegression()

# Pregunta #1 b):

**Entrena el modelo utilizando "engine-size" como la variable independiente y "price" como la variable dependiente.**

In [25]:
X = df[["engine-size"]]
Y = df[["price"]]

lm1.fit(X,Y)

LinearRegression()

# Pregunta #1 c):

**Encuentra la pendiente y la intersección del modelo.** 

#### Pendiente

In [26]:
# Slope
lm1.coef_

array([[166.86001569]])

#### Intersección

In [28]:
# Intercept
lm1.intercept_

array([-7963.33890628])

# Pregunta #1 d):

**¿Cuál es la ecuación de la línea predicha? Puedes usar x e yhat o "engine-size" o "price".**

In [29]:
# Usando X e Y

Yhat = -7963.34 + 166.86*X

Price = -7963.34 + 166.86*df["engine-size"]

¿Qué pasa si queremos predecir el precio de un automóvil usando más de una variable?

Si queremos usar más variables en nuestro modelo para predecir el precio de un automóvil, podemos usar la Regresión Lineal Múltiple. La Regresión Lineal Múltiple es muy similar a la Regresión Lineal Simple, pero este método se utiliza para explicar la relación entre una variable de respuesta continua (dependiente) y dos o más variables predictoras (independientes). La mayoría de los modelos de regresión del mundo real implican múltiples predictores. Ilustraremos la estructura utilizando cuatro variables predictoras, pero estos resultados pueden generalizarse a cualquier número entero.

$$
Y: Response \ Variable\\\\\\\\\\\\
X\_1 :Predictor\ Variable \ 1\\\\\\
X\_2: Predictor\ Variable \ 2\\\\\\
X\_3: Predictor\ Variable \ 3\\\\\\
X\_4: Predictor\ Variable \ 4\\\\\\
$$


$$
a: intercept\\\\\\\\\\\\
b\_1 :coeficientes \ de\ Variable \ 1\\\\\\
b\_2: coeficientes \ de\ Variable \ 2\\\\\\
b\_3: coeficientes \ de\ Variable \ 3\\\\\\
b\_4: coeficientes \ de\ Variable \ 4\\\\\\
$$


La ecuación es dada por:
$$
Yhat = a + b\_1 X\_1 + b\_2 X\_2 + b\_3 X\_3 + b\_4 X\_4
$$

De la sección anterior sabemos que otros buenos predictores del precio podrían ser:

- Caballos de fuerza
- Peso en vacío
- Tamaño del motor
- Millas por galón en carretera

Desarrollemos un modelo utilizando estas variables como las variables predictoras.

In [31]:
Z = df[["horsepower","curb-weight", "engine-size","highway-mpg"]]

Ajusta el modelo lineal utilizando las cuatro variables mencionadas anteriormente.


In [32]:
lm.fit(Z,df["price"])

LinearRegression()

¿Cual es el valor de la intersección (a)?

In [33]:
lm.intercept_

-15806.62462632923

¿Cuales son los valores de los coeficientes (b1,b2,b3,b4)?

In [34]:
lm.coef_

array([53.49574423,  4.70770099, 81.53026382, 36.05748882])